Code/Tools/AssetProcessor/native/resourcecompiler/rccontroller.cpp

```diff
@@ -334,22 +380,42 @@ namespace AssetProcessor
         if (!m_dispatchingJobs)
         {
             m_dispatchingJobs = true;
-            RCJob* rcJob = m_RCQueueSortModel.GetNextPendingJob();
 
-            while (m_RCJobListModel.jobsInFlight() < m_maxJobs && rcJob && !m_shuttingDown)
+            do
             {
-                if (m_dispatchingPaused)
+                RCJob* rcJob = m_RCQueueSortModel.GetNextPendingJob();
+
+                if (!rcJob)
                 {
-                    // note, even if dispatching is "paused" we start all "auto fail jobs" so that user gets instant feedback on failure.
-                    if (!rcJob->IsAutoFail())
+                    // there aren't any jobs remaining to dispatch.
+                    break;
+                }
+
+                // note that critical jobs and escalated jobs will always be at the top of the list
+                bool criticalOrEscalated = rcJob->IsCritical() || (rcJob->JobEscalation() > AssetProcessor::DefaultEscalation);
+
+                // do we have an open slot for this job?
+                unsigned int numJobsInFlight = m_RCJobListModel.jobsInFlight();
+                unsigned int regularJobLimit = m_alwaysUseMaxJobs ? m_maxJobs : AZStd::GetMax(m_maxJobs / 2, 1u);
+                unsigned int maxJobsToStart = criticalOrEscalated ? m_maxJobs : regularJobLimit;
+
+                // note that "auto fail jobs" oimmediately return as failed without doing any processing
+                // so they get to skip the line (they don't use up a thread
+                bool isAutoJob = rcJob->IsAutoFail();
+                bool tooManyJobs = numJobsInFlight >= maxJobsToStart;
+
+                if (!isAutoJob)
+                {
+                    if ((tooManyJobs) || (m_dispatchingPaused))
                     {
+                        // already using too much slots.
                         break;
                     }
                 }
                 StartJob(rcJob);
-                rcJob = m_RCQueueSortModel.GetNextPendingJob();
-            }
-            m_dispatchingJobs = false;
+            } while (true);
```

Comment:
Authors note: I'm not a fan of `do {} while(true)` but anything else in here would be redundant.
We want it to run thru this loop at least once, (so `do {} while`) and then break above for a various number of reasons.  So putting anything in the while (except true) would be checking something it already checks